feat(curriculum): job application form (#56396)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-job-application-form/66faac4139dbbd5dd632c860.md

---
id: 66faac4139dbbd5dd632c860
title: Build a Job Application Form
challengeType: 14
dashedName: lab-job-application-form
demoType: onClick
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should have a `div` element with the class `container`.
1. Inside the `div` element, you should have a `form` element.
1. The form should contain an `input` element with the type `text` and the id `name` for entering the user's full name.
1. You should have another `input` element with the type `email` and the id `email` for entering the user's email address.
1. The form should include a `select` element with the id `position` that allows users to select a job position.
1. You should have a `fieldset` element with class of `radio-group`.

> 1. Inside `.radio-group` you should have a set of `input` elements with the type `radio` for selecting availability options (e.g., Full-Time, Part-Time). The group `name` sshould be `availability`.

Reviewer: ```suggestion
1. Inside `.radio-group` you should have a set of `input` elements with the type `radio` and relevant labels for selecting availability options (e.g., Full-Time, Part-Time). The group `name` should be `availability`.
```

1. You should have a `textarea` element with the id `message` for entering a message.
1. You should have a `button` element with the type `submit` for submitting the form.
1. Add a `:focus` pseudo-class to the `input` and `textarea` elements to change their border color when focused.
1. The `input`, `select` and `textarea` elements should have an `:invalid` pseudo-class that changes the border color to red when invalid input is detected.
1. The `input`, `select` and `textarea` elements should have a `:valid` pseudo-class that changes the border color to green when valid input is entered.
1. The `button` element should have a `:hover` pseudo-class that changes the background color when hovered over.
1. Use the `:checked` pseudo-class on `.radio-group input[type="radio"]` to add a border color, background color and a box shadow when the radio button is selected.
1. Use the `:checked` pseudo-class on radio buttons to change the text color of the associated `label` when the option is selected.
1. Add an `:nth-child` pseudo-class to the `input` elements to style the first input fields differently. (e.g., rounded corners).

> 

Reviewer: Add the note for consistency across the labs:

```
**Note:** Be sure to link your stylesheet in your HTML and apply your CSS.
```

1. You should have a `textarea` element with the id `message` for entering a message.
1. You should have a `button` element with the type `submit` for submitting the form.
1. Add a `:focus` pseudo-class to the `input` and `textarea` elements to change their border color when focused.
1. The `input`, `select` and `textarea` elements should have an `:invalid` pseudo-class that changes the border color to red when invalid input is detected.
1. The `input`, `select` and `textarea` elements should have a `:valid` pseudo-class that changes the border color to green when valid input is entered.
1. The `button` element should have a `:hover` pseudo-class that changes the background color when hovered over.
1. Use the `:checked` pseudo-class on `.radio-group input[type="radio"]` to add a border color, background color and a box shadow when the radio button is selected.
1. Use the `:checked` pseudo-class on radio buttons to change the text color of the associated `label` when the option is selected.

> 1. The `button` element should have a `:disabled` pseudo-class that changes the background color when the form is incomplete or invalid.

Reviewer 2: Does it work without using JS?

Reviewer 3: I don't think it does

Reviewer 4: What parts need JS? From what I see in this section right here, it is completely possible with HTML and CSS.

Reviewer 3: `yeah, I'm removing it considering this is a html/css project

1. Add an `:nth-child` pseudo-class to the `input` elements to style the first input fields differently. (e.g., rounded corners).

> 1. Add an `:nth-child` pseudo-class to the `textarea` element to style the fifth textarea field differently. (e.g., border-bottom-left-radius).

Reviewer 2: I would completely remove this user story and related test as only one `textarea` has been required so far.

```js
assert.exists(document.querySelector("div.container > form select#position"));
assert.isNotEmpty(document.querySelectorAll("div.container > form select#position > option"))
```

You should have a `fieldset` or `section` element with the class `.radio-group`.

```js
assert.lengthOf(document.querySelectorAll('form fieldset.radio-group, form section.radio-group'), 1)
```

Inside `.radio-group` you should have a group of `input` elements with the type `radio` for selecting availability options. The group `name` should be `availability`.

```js
assert.isAtLeast(document.querySelectorAll("div.container > form .radio-group input[type='radio'][name='availability']").length, 2);
assert.isEmpty(document.querySelectorAll("div.container > form .radio-group input[type='radio']:not([name='availability'])"))
```

You should have a `textarea` element with the id `message` for entering a message.

```js
assert.isNotEmpty(document.querySelectorAll("div.container > form textarea#message"));
```

You should have a `button` element with the type `submit` for submitting the form.

```js
assert.isNotEmpty(document.querySelectorAll("div.container > form button[type='submit']"));
```

> Add a `:focus` pseudo-class to the `input` and `textarea` elements to change their border color when focused. Use a list selector in the given order.

Reviewer 2: ```suggestion
You should add a `:focus` pseudo-class to the `input` and `textarea` elements to change their border color when focused. Use a list selector in the given order.
```

```js
assert.isNotEmpty(new __helpers.CSSHelp(document).getStyle('input:focus, textarea:focus').getPropertyValue('border-color'))
```

The `input`, `select` and `textarea` elements should have an `:invalid` pseudo-class that changes the border color to red when invalid input is detected. Use a list selector in the given order.

```js
assert.equal(new __helpers.CSSHelp(document).getStyle('input:invalid, select:invalid, textarea:invalid').getPropertyValue('border-color'), 'red')
```

The `input`, `select` and `textarea` elements should have a `:valid` pseudo-class that changes the border color to green when valid input is entered. Use a list selector in the given order.

```js
assert.equal(new __helpers.CSSHelp(document).getStyle('input:valid, select:valid, textarea:valid').getPropertyValue('border-color'), 'green')
```

The `button` element should have a `:hover` pseudo-class that changes the background color when hovered over.

```js
assert.isNotEmpty(new __helpers.CSSHelp(document).getStyle('button:hover').getPropertyValue('background-color'))
```

> Use the `:checked` pseudo-class on `.radio-group input[type="radio"]` to add a border color, background color and a box shadow when the radio button is selected.

Reviewer 2: ```suggestion
You should use the `:checked` pseudo-class on `.radio-group input[type="radio"]` to add a border color, background color and a box shadow when the radio button is selected.
```

```js
const style = new __helpers.CSSHelp(document).getStyle('.radio-group input[type="radio"]:checked');
assert.isNotEmpty(style.getPropertyValue('border-color'));
assert.isNotEmpty(style.getPropertyValue('background-color'));
assert.isNotEmpty(style.getPropertyValue('box-shadow'));
```

> Use the `:checked` pseudo-class on radio buttons to change the text color of the associated `label` when the option is selected.

Reviewer 2: ```suggestion
You should use the `:checked` pseudo-class on radio buttons to change the text color of the associated `label` when the option is selected.
```
